sys/include/net/sock/dtls.h

```diff
@@ -0,0 +1,618 @@
+ * @{
+ */
+#define DTLS_CLIENT (1),    /**< Endpoint client role */
+#define DTLS_SERVER (2),    /**< Endpoint server role */
+/** @} */
+
+/**
+ * @brief Method of connecting to remote
+ */
+typedef struct {
+    /**
+     * @brief DTLS version number
+     *
+     * @see [DTLS version number](@ref sock_dtls_prot_version)
+     */
+    uint8_t dtls_version;
+
+    /**
+     * @brief DTLS role
+     *
+     * @see [DTLS role](@ref sock_dtls_role)
+     */
+    uint8_t role;
+} sock_dtls_method_t;
+
+/**
+ * @brief   Type for a DTLS sock object
+ *
+ * @note    API implementors: `struct sock_dtls` needs to be defined by
+ *          an implementation-specific `sock_dtls_types.h`.
+ */
+typedef struct sock_dtls sock_dtls_t;
+
+/**
+ * @brief Information about an established session with a remote endpoint.
+ */
+typedef struct sock_dtls_session sock_dtls_session_t;
+
+/**
+ * @brief Called exactly once during autoinit.
+ *
+ * Calls the initialization function required by the DTLS stack used.
+ */
+void sock_dtls_init(void);
+
+/**
+ * @brief Creates a new DTLS sock object
+ *
+ * Takes an initialized UDP sock and uses it for the transport.
+ * Memory allocation functions required by the underlying DTLS
+ * stack can be called in this function.
+ *
+ * @see net_credman.
+ *
+ * @param[out] sock     The resulting DTLS sock object
+ * @param[in] udp_sock  Existing UDP sock initialized with
+ *                      @ref sock_udp_create()to be used underneath.
+ * @param[in] tag       Credential tag of the sock. The sock will only use
+ *                      credentials with the same tag given here.
+ * @param[in] method    Defines the role of the endpoint and the DTLS version
+ *                      to use.
+ *
+ * @return  0 on success.
+ * @return  -1 on error
+ */
+int sock_dtls_create(sock_dtls_t *sock, sock_udp_t *udp_sock,
+                     credman_tag_t tag, sock_dtls_method_t method);
+
+/**
+ * @brief Initialises the server to listen for incoming connections
+ *
+ * @param[in] sock      DTLS sock to listen to
+ */
+void sock_dtls_init_server(sock_dtls_t *sock);
+
+/**
+ * @brief Establish DTLS session with a server.
+ *
+ * Initializes handshake process with a DTLS server @p ep.
+ *
+ * @param[in]  sock     DLTS sock to use
+ * @param[in]  ep       Endpoint to establish session with
+ * @param[out] remote   The established session, cannot be NULL
+ *
+ * @return  0 on success
+ * @return  -EAGAIN, if DTLS_HANDSHAKE_TIMEOUT is `0` and no data is available.
+ * @return  -EADDRNOTAVAIL, if the local endpoint of @p sock is not set.
+ * @return  -EINVAL, if @p remote is invalid or @p sock is not properly
+ *          initialized (or closed while sock_udp_recv() blocks).
+ * @return  -ENOBUFS, if buffer space is not large enough to store received
+ *          credentials.
+ * @return  -ETIMEDOUT, if timed out when trying to establish session.
+ */
+int sock_dtls_establish_session(sock_dtls_t *sock, sock_udp_ep_t *ep,
```

Comment:
BTW. https://github.com/RIOT-OS/RIOT/pull/10308 uses the wording `session_create`/`session_destroy` for these functions. It's non-blocking, but I think those names are a bit more succinct.

Reply:
renamed.